flipperdevices/flipperzero-firmware
# [FL-3762] Configurable Infrared TX output

# What's new

- Add menu to configure:
  - Infrared transmit pin (built-in LED, PA7 or Detect) (saved between application launches)
  - 5V supply at the GPIO header (resets on application exit)
- Automatic module detection (works for modules that use a FET with a strong pull-down or a BJT as the LED driver)

# Verification 

1. Flash the firmware and launch the app (the latter needs API version 58.1).
2. Go to `GPIO Settings` and set `Signal Output` to `2 (A7)`.
3. Connect an infrared LED to the pin 2 (do not forget to include an appropriate series resistor!).
4. Try transmitting an infrared signal (learn it or load a saved one). Check that a receiver (another Flipper or something else) can receive and make sense of it.
5. Check that no signal is transmitted via the built-in LEDs. (use a phone camera for example).
6. Exit the application, relaunch it. Go to `Gpio Settings` and ensure that the `Signal Output` setting still says `2 (A7)`.
7. Set it back to `Flipper`. Now Flipper should operate in standard mode. Run the usual Infrared checks to ensure that it is really so.
8. Disconnect Flipper from USB power, if it is currently connected.
9. Go to `GPIO Settings` and set `5V on GPIO` to `ON`.
10. Measure the voltage at the GPIO pin 1 and check that 5V is indeed present on it.
11. Set `5V on GPIO` to `OFF` and then back to `ON`. Measure each time to ensure that the voltage output follows the setting.
12. Without exiting the app, go into its various menus. The 5V voltage should remain present at all times.
13. Exit the application. The 5V voltage should disappear from pin 1.
14. Relaunch the application. The `5V on GPIO` setting should be reset back to `OFF` and no voltage sould be present at GPIO pin 1.
15. Connect an IR amplifier module (e.g. IR Blaster) to Flipper's GPIO.
16. Set `GPIO Settings -> Signal Output` to `Auto`. The 5V power should become enabled right away.
17. Repeat steps 4-5. The infrared signal should be transmitted via the module.
18. Disconnect the module and relaunch the application without changing any settings.
19. Repeat step 3. The infrared signal should be transmited via the built-in LEDs and no 5V power should be enabled.

# Checklist (For Reviewer)
